### PR TITLE
KAFKA-6554; Missing lastOffsetDelta validation before log append

### PR DESCRIPTION
**Pull request by hachikuji**
Add validation checks that the offset range is valid and aligned with the batch count prior to appending to the log. I've added several unit tests to verify the various invalid cases.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by hachikuji** [comment]
cc @rajinisivaram

File: core/src/main/scala/kafka/log/LogValidator.scala [hunk 1]
```diff
@@ -74,15 +74,26 @@ private[kafka] object LogValidator extends Logging {
...
+        if (countFromOffsets <= 0)
+          throw new InvalidRecordException("Batch has an invalid offset range")
+
+        // v2 and above messages always have a non-null count
+        val count = batch.countOrNull
+        if (count <= 0)
+          throw new InvalidRecordException("Invalid reported count for record batch")
+
+        if (countFromOffsets != batch.countOrNull)
+          throw new InvalidRecordException("Inconsistent batch offset range and count of records")
```
Comment (rajinisivaram) [inline core/src/main/scala/kafka/log/LogValidator.scala]: Would it be useful to include the two counts in the exception message?

File: core/src/test/scala/unit/kafka/log/LogValidatorTest.scala [hunk 2]
```diff
@@ -147,6 +148,43 @@ class LogValidatorTest {
...
+    validateRecordBatchWithCountOverrides(lastOffsetDelta = 15, count = 3)
+    validateRecordBatchWithCountOverrides(lastOffsetDelta = -3, count = 3)
+    validateRecordBatchWithCountOverrides(lastOffsetDelta = 2, count = -3)
+    validateRecordBatchWithCountOverrides(lastOffsetDelta = 2, count = 6)
+    validateRecordBatchWithCountOverrides(lastOffsetDelta = 2, count = 0)
+    validateRecordBatchWithCountOverrides(lastOffsetDelta = -3, count = -2)
+
+    // Count and offset range are consistent with each other, but do not match the actual number of records
+    validateRecordBatchWithCountOverrides(lastOffsetDelta = 5, count = 6)
+    validateRecordBatchWithCountOverrides(lastOffsetDelta = 1, count = 2)
```
Comment (rajinisivaram) [inline core/src/test/scala/unit/kafka/log/LogValidatorTest.scala]: Do we also want a test here with `lastOffsetDelta = 2, count = 3` that doesn't throw an exception (for completion even though it is tested elsewhere)?

**Review by rajinisivaram — APPROVED** [review]
@hachikuji Thanks for the PR. Left a couple of minor comments. LGTM (feel free to merge, doesn't require another review).

**Comment by rajinisivaram** [comment]
@hachikuji Thanks for the updates, LGTM. Merging to trunk and 1.1.

**Pull request merged**